executor/infoschema_reader.go

```diff
@@ -2407,7 +2412,25 @@ func (r *dataLockWaitsTableRetriever) retrieve(ctx context.Context, sctx session
 
 		// Calculate rows.
 		res = make([][]types.Datum, 0, end-start)
-		for rowIdx, lockWait := range r.lockWaits[start:end] {
+		var lockWaitsStart int
+		var lockWaitsEnd int
+		var resolvingStart int
+		var resolvingEnd int
+		if start < len(r.lockWaits) {
+			lockWaitsStart = start
+			resolvingStart = 0
+		} else {
+			lockWaitsStart = len(r.lockWaits)
+			resolvingStart = start - len(r.lockWaits)
+		}
+		if end < len(r.lockWaits) {
+			lockWaitsEnd = end
+			resolvingEnd = 0
+		} else {
+			lockWaitsEnd = len(r.lockWaits)
+			resolvingEnd = end - len(r.lockWaits)
+		}
```

Comment:
It can be simplified as:
```suggestion
        lockWaitsStart := mathutil.Min(start, len(r.lockWaits))
		resolvingStart := start - lockWaitsStart
		lockWaitsEnd := mathutil.Min(end, len(r.lockWaits))
		resolvingEnd := end - lockWaitsEnd
```
And I suggest adding some comments to illustrate it.